Update cs50.cpp

IvanJasenov opened the pull request:
Convert std::string type string into c_str(), NULL terminated string.
The rest of the get_int() method works the same as within lib50-c both in C++98 and C++11.

IvanJasenov committed:
Update cs50.cpp

Convert std::string type string into c_str(), NULL terminated string.
The rest of the get_int() method works the same as within lib50-c

AndrejPetelin commented on src/cs50.cpp:
```diff
@@ -57,16 +74,29 @@ long long get_long_long(void)
+	}
+	// do the conversion c_str
+	static int len = s.length();
+	char* str = new char(len + 1);
```
This is a memory leak, isn't it?

IvanJasenov replied:
Yes, it's leaking! Not a final version, just a draft one. Looking forward for whether this idea can pass the bar or not. Garbage collector is something I'd add at the very end, Mark and Sweep.

AndrejPetelin replied:
What do you mean by garbage collector? There are garbage collectors for C++ but in general the idea is to use RAII - std::string handles its own memory. What you're doing here is returning a C-style string, allocated on free-store via a `std::string`.  If you're reading into a `std::string`, returning an `std::string`, what's the point of dynamically allocating a char array in the first place? And what happens on return is probably not what you think is happening. You're passing a char array to a `std::string`. `std::string` has a constructor that takes a char array as argument and copies the contents. Hence the memory leak, and needlessly so.

IvanJasenov replied:
I needed something which will be NULL terminated and std::strings are not. And for such format, I do need GC, eventually.

irenenaya replied:
But what null terminated are we talking about? the function returns an std::string. That's not a NULL terminated string.

AndrejPetelin replied:
If you need GC you're probably not using the correct language. C++ generally prefers RAII (allocate in constructor, deallocate in destructor, no "naked" `new`s. Why do you need a null terminated string? std::string has `size()` and `length()` that return exactly that.

AndrejPetelin commented on src/cs50.cpp:
```diff
@@ -57,16 +74,29 @@ long long get_long_long(void)
+		throw std::invalid_argument("EOF");
+	}
+	// do the conversion c_str
+	static int len = s.length();
```
Why static? Isn't the length of string going to change among function calls? Why do you need len in the first place? `s.length()` is constant time anyway.

IvanJasenov replied:
I think it should not change, since other methods use get_string as a parser.

AndrejPetelin replied:
User enters:

>Simon
Garfunkel
Hello darkness, my old friend, I've come to talk with you again

len should be 5 for all 3 entries?

irenenaya replied:
Yes, it will be 5 for all three entries. Only that for the 3rd one the program will 1) crash with a memory error and 2) the string will not be correct because strcpy will likely not copy it correctly because the size of the string is still 5 and the input string is much larger

AndrejPetelin commented on src/cs50.cpp:
```diff
@@ -57,16 +74,29 @@ long long get_long_long(void)
+	{
+		throw std::invalid_argument("Something went wrong");
+	}
+	else if(std::cin.eof()) // if not enough memory
```
eof is end of file, not out of memory. Plus is there a point, you're throwing invalid_argument when cin is not in good state anyway.

irenenaya commented:
You haven't even tried compiling this, have you? get_int() has an implicit conversion from std::string to char*, so it won't compile.

IvanJasenov commented:
@irenenaya @AndrejPetelin My bad, have 6 draft version in front of me and the ones I've tested on have char* as return type of get_string(). This one slips under the radar.

IvanJasenov committed:
Update cs50.cpp

I suggest that the return type of get_string to be char* type so it wont raise and error if something like, std::string s = get_string() is invoked later on.

IvanJasenov commented:
I'll be back!
